### ISSUE TITLE
Running Exoplayer in a service connected to a PlayerView on UI for AnalyticsListener and VideoListener features

### ISSUE THREAD

**Issue by ilker-aktuna** [google/ExoPlayer#8071]
Checked Stackoverflow and GitHub issues but could not address my issue directly

I am building an app to analyze live TV video streams and report when a drop, loaderror or other issue is detected.
So the app will run on a low profile Android phone (current test phone: Samsung J7 , 1GB  RAM , Android 7.1 OS)
The phone will run Exoplayer in a service and open a different HLS stream every N seconds (15 seconds at the moment)
Then using the AnalyticsListener and VideoListener features it will collect information and events from streams.

When I tried this, I see that VideoListener does not work if the player is not bound to a PlayerView which is visible.
Is this correct ?
If there is a way to use VideoListener without any visible PlayerView, can you tell me ?

In the worst case, I force the app to leave screen on and disable keyguard, it works.
But I have this problem:

Since the player is using the playerview on UI thread I have to release and start an new player every 15 seconds.
I use this code:

main service thread:

```
                    MyService.runOnUI(new Runnable() {
                        public void run() {
                            if (player.getPlaybackState() == Player.STATE_BUFFERING) {
                                //player.release();
                                //startPlayer();
                                reStartPlayer();
                            } else {
                                startPlayer();
                            }
                        }
                    });

                    Handler handler = new Handler(Looper.getMainLooper());
                    Runnable runnableCode = new Runnable() {
                        @Override
                        public void run() {
                            postStat();
                        }
                    };
                    handler.postDelayed(runnableCode, 15000);
```

so the main thread calls restartPlayer() and then after 15 seconds it uses postStat() function to post collected stats, and goes back to start.

restartPlayer function is like:

```
 public void reStartPlayer() {
        player.release();
        try {
            Thread.sleep(500);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        player.removeAnalyticsListener(playerAnalyticsListener);
        player.removeVideoListener(playerVideoListener);
        player=null;
        DefaultLoadControl loadControl = new DefaultLoadControl.Builder().setPrioritizeTimeOverSizeThresholds(false).setBufferDurationsMs(2500, 10000, 2500, 2500).build();
        player = new SimpleExoPlayer.Builder(getApplicationContext()).setLoadControl(loadControl).build();
        player.setVolume(0);
        playerAnalyticsListener=new MainAlayticsListenerForService(player);
        player.addAnalyticsListener(playerAnalyticsListener);
        player.addVideoListener(playerVideoListener);
        playerView = myPlayerView();
        playerView.setPlayer(player);
        playerView.requestFocus();

        Uri hlsVideoUri = Uri.parse(currURL);
        DataSource.Factory dataSourceFactory = new DefaultDataSourceFactory(this,
                Util.getUserAgent(this, "Test"));
        HlsMediaSource hlsMediaSource = new HlsMediaSource.Factory(dataSourceFactory).createMediaSource(hlsVideoUri);
        player.setPlayWhenReady(true);
        player.prepare(hlsMediaSource);
    }
```

Now this works, but after some time of running , it stops with following exception:

```
2020-10-13 12:05:03.339 26492-26513/? W/OpenGLRenderer: swapBuffers encountered EGL error 12301 on 0x9b5cd0e0, halting rendering...
2020-10-13 12:05:03.339 26492-31666/? E/art: ashmem_create_region failed for 'indirect ref table': Too many open files
2020-10-13 12:05:03.340 26492-31666/? W/art: Throwing OutOfMemoryError "Could not allocate JNI Env"
2020-10-13 12:05:03.348 26492-31666/? E/ExoPlayerImplInternal: Playback error
      com.google.android.exoplayer2.ExoPlaybackException: Unexpected runtime error
        at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:564)
        at android.os.Handler.dispatchMessage(Handler.java:98)
        at android.os.Looper.loop(Looper.java:154)
        at android.os.HandlerThread.run(HandlerThread.java:61)
     Caused by: java.lang.OutOfMemoryError: Could not allocate JNI Env
        at java.lang.Thread.nativeCreate(Native Method)
        at java.lang.Thread.start(Thread.java:730)
        at com.google.android.exoplayer2.audio.DefaultAudioSink.flush(DefaultAudioSink.java:1167)
        at com.google.android.exoplayer2.audio.MediaCodecAudioRenderer.onDisabled(MediaCodecAudioRenderer.java:511)
        at com.google.android.exoplayer2.BaseRenderer.disable(BaseRenderer.java:175)
        at com.google.android.exoplayer2.ExoPlayerImplInternal.disableRenderer(ExoPlayerImplInternal.java:1510)
        at com.google.android.exoplayer2.ExoPlayerImplInternal.seekToPeriodPosition(ExoPlayerImplInternal.java:1131)
        at com.google.android.exoplayer2.ExoPlayerImplInternal.seekToPeriodPosition(ExoPlayerImplInternal.java:1095)
        at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMediaSourceListInfoRefreshed(ExoPlayerImplInternal.java:1690)
        at com.google.android.exoplayer2.ExoPlayerImplInternal.setMediaItemsInternal(ExoPlayerImplInternal.java:670)
        at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:505)
        at android.os.Handler.dispatchMessage(Handler.java:98) 
        at android.os.Looper.loop(Looper.java:154) 
        at android.os.HandlerThread.run(HandlerThread.java:61) 
2020-10-13 12:05:03.410 26492-26503/? I/art: Background sticky concurrent mark sweep GC freed 109466(10MB) AllocSpace objects, 10(1856KB) LOS objects, 8% free, 108MB/118MB, paused 22.656ms total 231.827ms
```

I don't understand why I get " Too many open files" error because I release the player every 15 seconds before creating a new one.
What can I do about this ?

thanks.

**Comment by marcbaechinger** [comment]
When you are in a service and the `PlayerView` is not in a view hierarchy built up by an activity/fragment you can not do a `findById(id)` to get the view. How do you create that view that you return in `myPlayerView`?

```
playerView = myPlayerView();
```

Is the log exception the only thing you see in the logs? I wonder whether you can add an `EventLogger` to the app, repro again and send me a bug report/logcat output. We then see whether a) the video track is disabled and b) whether there is some other things the player complains about.  I actually thing that somehow the video codec renderer is disabled and hence you would not see events in the `VideoListener`.

```
player.addAnalyticsListener(new EventLogger(/* trackSelector= */null));
```

**Comment by ilker-aktuna** [comment]
myPlayerView is my function under activity which returns the playerview from that activity:

```
    public static PlayerView myPlayerView() {
        return playerView;
    }
```

Actually I don't use a trackselector. I use a Uri to build a HlsMediaSource:

```
        DataSource.Factory dataSourceFactory = new DefaultDataSourceFactory(this, Util.getUserAgent(this, "Test"));
        HlsMediaSource hlsMediaSource = new HlsMediaSource.Factory(dataSourceFactory).createMediaSource(hlsVideoUri);
```

If you tell me how to create the eventlogger, I can get you the logs with events that you expect.

**Comment by marcbaechinger** [comment]
Does this not work as above?

```
player.addAnalyticsListener(new EventLogger(/* trackSelector= */ null));
```

**Comment by ilker-aktuna** [comment]
I didn't understand how to implement that.
I already have an analyticslistener created from a custom class.

```
                playerAnalyticsListener=new MainAlayticsListenerForService(player);
                player.addAnalyticsListener(playerAnalyticsListener);

```

do you want me to add another analyticslistener ?
and what shall I observe ?

thanks.

**Comment by marcbaechinger** [comment]
Ah, ok, sorry. Yes, you can add as many analytics listener as you want. `EventLogger` just prints a nice log statement to the console for each event it receives. So you logcat will have a pretty nice output to see what is going on in the player.

If you want to interpret yourself, there is a[ section in the developer guide](https://exoplayer.dev/listening-to-player-events.html#using-eventlogger) about it. I will look into this as well when you send the output over.

**Comment by ilker-aktuna** [comment]
ok. I will do as you suggested and get a log when the problem is reproduced.

But I want to ask; on which part of the problem are we trying to focus on right now ?
The exception says " Too many open files" . Shall we be able to see what  causes this through the AnalyticsListener ?

**Comment by marcbaechinger** [comment]
I'm not sure what is not working. I would expect that there is something in the logs that gives us more information where to dig.
At least I can see whether the player is initialized, prepared and destroyed as expected. If that is not the case we can investigate some further to find whether some components do not close files properly or similar. Short: we want to check whether the too many open files/Out of Memory is a side effect of another problem. Does that make sense?

**Comment by ilker-aktuna** [comment]
sure. I'll try get it.

**Comment by ilker-aktuna** [comment]
I've added  the code for eventlogger and now run the app. Usually  the memory problem occurs after an hour of continuous run.
I can share the current events if you want. Normally it goes through the same process every 15 seconds.
How should I send the log ?

**Comment by marcbaechinger** [comment]
You can upload here or if you're unable to share bug reports or test content publicly, please send them to dev.exoplayer@gmail.com using a subject in the format "Issue #8071". Please also update this issue to indicate you’ve done this.

**Comment by ilker-aktuna** [comment]
in the attached file I have events from the running process.
The app has not crashed yet, so I don't have the exception with "too many open files"  issue.
When I catch it I will also upload this. But meanwhile maybe you can catch something that I'm doing wrong which causes unreleased resources.

Also, is there a way to use "videolistener" without keeping the display on and  playing video ?
I am using videolistener to get bitrate and resolution of the video.
I am using analyticslistener to catch "loadcomplete, loaderror, drop, discontinuity, playererror" events.
But analyticslistener works without an attached viewer. So , the thing that force me to use a visible playerview is actually bitrate and resolution.
[exoplayer.zip](https://github.com/google/ExoPlayer/files/5378211/exoplayer.zip)

**Comment by marcbaechinger** [comment]
Thanks for the logs.

After looking into this I think that something is not going on as we expect it from your report. 

1) Init/Release: When ExoPlayer is [inititalized](https://github.com/google/ExoPlayer/blob/release-v2/library/core/src/main/java/com/google/android/exoplayer2/ExoPlayerImpl.java#L136) and [released](https://github.com/google/ExoPlayer/blob/release-v2/library/core/src/main/java/com/google/android/exoplayer2/ExoPlayerImpl.java#L705)  a log line is printed to the console. So if you create a new player every 15 seconds as in the code sample you posted above, I would expect such a line every 15 seconds in the log, but I can't find such a line. It seems that the player is created only once and then recreated. 

The logs start with 

```
timeline [eventTime=6.53, mediaPos=86362.32, window=0, period=0, periodCount=1, windowCount=1, reason=SOURCE_UPDATE
```

So the Init line is probably just before the log I have. However, there should be a line with `Release ` every 15 seconds which is missing.

2)  VideoListener log lines

In the log I find lines as the ones below:

```
EventLogger: videoSize [eventTime=28.91, mediaPos=86330.04, window=0, period=0, 1280, 720]
EventLogger: renderedFirstFrame [eventTime=28.91, mediaPos=86330.04, window=0, period=0, Surface(name=null)/@0xedc9b40]
```

These are lines of the video listener. So I actually would expect that the video listener is called as expected. At least these calls are printed by the `EventListener` according to the log file you sent.


It's unclear why your `VideoListener` is not called while the calls arrive at the `EventLogger`. You can solve this by getting resolution from the analytics listener `MainAlayticsListenerForService` you are already using. It has the same callbacks than the video listener. This probably solves your problem, but probably doesn't explain the behaviour you are seeing.

Further it seems that something is not working as expected according to the code you posted above (missing Init/Release lines in the logfile). 

With that evidence, I don't see that there is something wrong with the player. Can you review your setup and see whether everything works as you intended? Live can you double-check whether the code above is still the same and if it is actually executed?

**Comment by ilker-aktuna** [comment]
thanks for analyzing it.
actually there is only one place in my code that I change the source. and that is where I call the restartPlayer() function.
it is called every 15 seconds and the code for the funtion is :

```
 public void reStartPlayer() {
        player.release();
        try {
            Thread.sleep(500);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        player.removeAnalyticsListener(playerAnalyticsListener);
        player.removeVideoListener(playerVideoListener);
        player=null;
        DefaultAllocator allocator = new DefaultAllocator(true, C.DEFAULT_BUFFER_SEGMENT_SIZE);
        DefaultLoadControl loadControl = new DefaultLoadControl.Builder().setPrioritizeTimeOverSizeThresholds(false).setBufferDurationsMs(2500, 10000, 2500, 2500).build();
        player = new SimpleExoPlayer.Builder(getApplicationContext()).setLoadControl(loadControl).build();
        player.setVolume(0);
        playerAnalyticsListener=new MainAlayticsListenerForService(player);
        player.addAnalyticsListener(playerAnalyticsListener);
        player.addAnalyticsListener(new EventLogger(/* trackSelector= */ null));
        player.addVideoListener(playerVideoListener);
        playerView = myPlayerView();
        playerView.setPlayer(player);
        playerView.requestFocus();

        Uri hlsVideoUri = Uri.parse(currURL);
        DataSource.Factory dataSourceFactory = new DefaultDataSourceFactory(this,
                Util.getUserAgent(this, "VTV Test"));
        HlsMediaSource hlsMediaSource = new HlsMediaSource.Factory(dataSourceFactory).createMediaSource(hlsVideoUri);
        player.setPlayWhenReady(true);
        player.prepare(hlsMediaSource);
    }
```

so as you see it has the release function. it also has "player=null" to make sure we create a new player every time.
how come it is not logged in the events as you can't see ?

btw, I could not understand your comment:
"It's unclear why your VideoListener is not called while the calls arrive at the EventLogger"

my video listener is called and I can get what I need (bitrate and resolution) from it. my problem is not about videolistener not being called. my problem is that I need to use videolistener with a visible playerview to get what I need.
and my question is :
is there a way to use it with no visible playerview ?

and you say that same events for getting resolution & bitrate are available in analyticslistener.
if that's correct, then I can remove the videolistener completely.
but will it let me get the bitrate and resolution without a visible playerview ?

also, my most important problem is  the memory and "too many open files" exception.
how shall I troubleshoot that ?

thanks.

**Comment by marcbaechinger** [comment]
> is there a way to use it with no visible playerview ?

I think the `MediaCodecVideoRenderer` needs a surface to do the rendering. So you need to provide your surface. Doing this without a `PlayerView` is an advanced topic which I probably can't support you with I'm afraid.

> also, my most important problem is the memory and "too many open files" exception.

Yes, that's why I wanted to see the logs and analyzed it. I actually don't think this is a player problem, but an app problem. I haven't found anything suspicious except that it seems to me that the player is not released as your code suggests.

In general, it should not be required to create new player. I'd try just to prepare again. Please also look [at the demo app to see how the player is created once, while it can be prepared multiple times](https://github.com/google/ExoPlayer/blob/release-v2/demos/main/src/main/java/com/google/android/exoplayer2/demo/PlayerActivity.java#L276).

Both methods should work.

**Comment by ilker-aktuna** [comment]
ok. I actually tested now without a videolistener and just before changing the source, I can get bitrate and resolution with : 
```
                            Format vF = player.getVideoFormat();
                            if(vF!=null) {
                                bitRate = vF.bitrate;
                                chHeight = vF.height;
                                chWidth = vF.width;
                            }

```

so I will stop using the videolistener.
and that will let me use the app with no obligation to let screen on.
that is good. 

for the "too many open files" issue , I will check further. let me try to get the exact moment where that occurs.

but you say that I don't need to release the player for changing the source ?

**Comment by ilker-aktuna** [comment]
I removed the videolistener and now getting the bitrate/resolution from videoFormat as soon as the video is loaded. 
now the app does not crash, or produce memory issues.
maybe it solved the whole issue

**Comment by marcbaechinger** [comment]
When you do this

```
MyService.runOnUI(new Runnable() {
                        public void run() {
                            if (player.getPlaybackState() == Player.STATE_BUFFERING) {
                                //player.release();
                                //startPlayer();
                                reStartPlayer();
                            } else {
                                startPlayer();
                            }
                        }
                    });
```

the player is restarted only when the player is buffering. In all other cases you call startPlayer(). I assume you do not release properly in startPlayer which results in many players being created which eventually gives you the too many open files/Out of memory problem.

Please review your code to make sure to create and release the player properly. Study the methods `initializePlayer` and `releasePlayer`  in the demo app as an example. With the EventLogger you can check what is going on.

**Comment by ilker-aktuna** [comment]
well, the startPlayer is only called when the player is not buffering. doesn't that also mean , "when player is not playing"
what is recommended to change the stream ?
should I release and recreate ?
should I set it to null and recreate ?
or should I just prepare ? (no release) 

I know you recommend to check the demo app but it is far more complicated then my app.
I don't use a trackselector , instead I change the Uri for HLS and prepare the player.
so I just need to know a simple way of just changing the source.

**Comment by ilker-aktuna** [comment]
after removing videolistener (my original impl. and the one for eventlogger),  and removing unnecessary runOnUI calls, now the app does not crash frequently.
So I don't observe "too many open files" or "out of memory" issues. If I can catch one, I will report back.

On the other hand, I still did not understand what is the correct method for changing source from one Uri to another (HLS)
Should I release the player and recreate it ?
Should I just change Uri and "prepare" ?

**Comment by marcbaechinger** [comment]
There is not need to release the player or creating a new instance of it. 

With the new API released with 2.12 you can do this by just using [`setMediaItem(item)`](https://github.com/google/ExoPlayer/blob/release-v2/library/core/src/main/java/com/google/android/exoplayer2/Player.java#L888) again:

```
player.setMediaItem(mediaItem);
player.prepare();

// later to set another source
player.setMediaItem(newMediaItem);
```

If you are using the `prepare(MediaSource)` method that is deprecated since 2.12 I would do:

```
exoPlayer.prepare(mediaSource)

// later for preparing with another source
exoPlayer.prepare(newMediaSource)
```

**Comment by ilker-aktuna** [comment]
ok thanks. I removed the release calls. Now it performs better.
I have another question which is off-topic.
Where should I normally ask questions about using some features like analytics listener ?

actually, I am using the analytics listener to see errors when loading the stream.
so I count loadcomplete and loaderror events.
But still chunks are variable according to the available bandwidth and size of video/time, I never observe a static count of loadcomplete events. So I can't compare what I observe. 
What should I do to get a static count of loadcomplete events for a predefined duration of stream ?

**Comment by marcbaechinger** [comment]
This issue tracker is the right place to ask. Best is to create a new issue for this so the title matches the question. This way other user who have a similar question can find it and it's not off-topic in this issue. 

If you open a new issue, it's helpful to also describe quickly what your use case is. Because then we can probably advice you which callback of the `AnalyticsListener` is the best to use.

**Comment by ilker-aktuna** [comment]
thanks. I've opened a new issue for it.
And , now I will close this one if you've nothing else to add ?
